File: LC/Python/Hard/124.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = [-inf]
        
        def dfs(curr):
            if not curr:
                return 0
            
            left = dfs(curr.left)
            right = dfs(curr.right)
            
            if not right or right < 0:
                right = 0
            if not left or left < 0:
                left = 0
            
            curr_sum = curr.val + left + right
            ans[0] = max(ans[0], curr_sum)
            
            return max(right, left) + curr.val
        
        dfs(root)
        
        return ans[0]

File: LC/C++/Hard/124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int ans_ = -INT_MAX;
public:
    int dfs(TreeNode* curr) {
        if (curr == nullptr) { return 0; }

        int left = curr->left == nullptr ? 0 : dfs(curr->left);
        if (left < 0) {
            left = 0;
        }
        
        int right = curr->right == nullptr ? 0 : dfs(curr->right);
        if (right < 0) {
            right = 0;
        }

        int res = left + right + curr->val;
        ans_ = max(ans_, res);
        return max(left, right) + curr->val;
    }

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return ans_;
    }
};